ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -174,9 +174,220 @@ static void *submsg_handler(void *closure, const void *hd) {
   return submsg;
 }
 
+// Handler data for startmap/endmap handlers.
+typedef struct {
+  size_t ofs;
+  upb_fieldtype_t key_field_type;
+  upb_fieldtype_t value_field_type;
+  VALUE value_field_typeclass;
```

Comment:
I think that we actually do need `typeclass_references` -- the reason is that each `Descriptor` is rooted separately in the def --> Ruby object global map, and a given descriptor refers to its fields' typeclasses only on the upb side, not on the Ruby side. For now we would be fine because the global map is never trimmed, but if in the future we turn it into a true cache with eviction, we would be in trouble. (RepeatedField and singular fields don't run into this issue because they dynamically look up the typeclass when they need it via that global map.)